File: notification_service/src/infra/redis/redis.ts
import { Redis } from "ioredis";
import { RedisConfig } from "../../config/index.js";
import { logger } from "../logger/index.js";

let redisInstance: Redis | null = null;

export function getRedisClient(): Redis {
  if (!redisInstance) {
    redisInstance = new Redis(RedisConfig.REDIS_URL,{
      maxRetriesPerRequest: null,
    });

    redisInstance.on("connect", () => {
      logger.info("✅ Redis connected");
    });

    redisInstance.on("error", (err) => {
      logger.error("❌ Redis error:", err);
    });
  }

  return redisInstance;
}

File: api_gateway_service/config/redis.go
package config

import (
	"context"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
)

func LoadRedis() (*redis.Client, error) {
	otps, err := redis.ParseURL(RedisConfig.REDIS_URL)
	if err != nil {
		return nil, fmt.Errorf("failed to parse redis url: %v", err)
	}

	rclient := redis.NewClient(otps)

	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := rclient.Ping(ctx).Err(); err != nil {
		return nil, fmt.Errorf("failed to ping redis: %v", err)
	}

	fmt.Println("Successfully connected to redis")

	return rclient, nil
}
